src/src/PluginStructs/P104_data_struct.cpp

```diff
@@ -0,0 +1,2458 @@
+  if (!initialized) {
+    loadSettings();
+    initialized = true;
+  }
+
+  if ((P != nullptr) && (cs_pin > -1)) {
+    addLog(LOG_LEVEL_INFO, F("dotmatrix: P->begin() called"));
+    P->begin(expectedZones);
+    # ifdef P104_USE_BAR_GRAPH
+    pM = P->getGraphicObject();
+    # endif // ifdef P104_USE_BAR_GRAPH
+    return true;
+  }
+  return false;
+}
+
+# define P104_ZONE_SEP   '\x02'
+# define P104_FIELD_SEP  '\x01'
+# define P104_ZONE_DISP  ';'
+# define P104_FIELD_DISP ','
+
+# define P104_CONFIG_VERSION_V2  0xF000 // Marker in first uint16_t to to indicate second version config settings, anything else if first
+                                        // version.
+                                        // Any third version or later could use 0xE000, etc. The 'version' is stored in the first uint16_t
+                                        // stored in the custom settings
+
+/*
+   Settings layout:
+   Version 1:
+   - uint16_t : size of the next blob holding all settings
+   - char[x]  : Blob with settings, with csv-like strings, using P104_FIELD_SEP and P104_ZONE_SEP separators
+   Version 2:
+   - uint16_t : marker with cpntent P104_CONFIG_VERSION_V2
+   - uint16_t : size of next blob holding 1 zone settings
+   - char[y]  : Blob holding 1 zone settings, with csv like string, using P104_FIELD_SEP separators
+   - uint16_t : next size, if 0 then no more blobs
+   - char[x]  : Blob
+   - ...
+   - Max. allowed total custom settings size = 1024
+ */
+/**************************************
+ * loadSettings
+ *************************************/
+void P104_data_struct::loadSettings() {
+  uint16_t bufferSize;
+  char    *settingsBuffer;
+
+  if (taskIndex < TASKS_MAX) {
+    int loadOffset = 0;
+
+    // Read size of the used buffer, could be the settings-version marker
+    LoadFromFile(SettingsType::Enum::CustomTaskSettings_Type, taskIndex, (uint8_t *)&bufferSize, sizeof(bufferSize), loadOffset);
+    bool settingsVersionV2  = (bufferSize == P104_CONFIG_VERSION_V2) || (bufferSize == 0u);
+    uint16_t structDataSize = 0;
+    uint16_t reservedBuffer = 0;
+
+    if (!settingsVersionV2) {
+      reservedBuffer = bufferSize + 1;              // just add 1 for storing a string-terminator
+      addLog(LOG_LEVEL_INFO, F("DOTMATRIX: Reading Settings V1, will be stored as Settings V2."));
+    } else {
+      reservedBuffer = P104_SETTINGS_BUFFER_V2 + 1; // just add 1 for storing a string-terminator
+    }
+    reservedBuffer++;                               // Add 1 for 0..size use
+    settingsBuffer = new char[reservedBuffer]();    // Allocate buffer and reset to all zeroes
+    loadOffset    += sizeof(bufferSize);
+
+    if (!settingsVersionV2) {
+      structDataSize = bufferSize;
+    } else {
+      LoadFromFile(SettingsType::Enum::CustomTaskSettings_Type, taskIndex, (uint8_t *)&bufferSize, sizeof(bufferSize), loadOffset);
+      structDataSize = bufferSize;
+      loadOffset    += sizeof(bufferSize); // Skip the size
```

Comment:
Is this correct? You offset it with `sizeof(bufferSize)` (=2) twice.

Reply:
It has changed quite a few times since I initially wrote it, so it is now in a 'carefully crafted' situation, I guess 😆
I'm trying to backtrack why it ended up this way, but AFAICS it is working as intended. Most reasonable explanation is that it has to do with switching from saving as 1 big blob to a per-zone, dynamic size, save/load strategy.

Reply:
Aha, the V2 settings have a double `bufferSize` at the start of the storage, that is now used as an indicator for V2. Then there is the actual `bufferSize` of the first saved block and the save data of that length.